vehicle.py

```diff
@@ -0,0 +1,262 @@
+        if is_in_base:
+            self.priority = None
+            return
+        our_tanks_in_base = state.get_our_tanks_cells().intersection(map_.base)
+        empty_base_cells = map_.base.difference(state.tank_cells)
+        if empty_base_cells and len(our_tanks_in_base) < 2:
+            self.priority = empty_base_cells.pop()
+        if self.priority is None or self.priority == self.model.coordinates:
+            free_cells = map_.available_cells.difference(state.tank_cells)
+            self.priority = random.choice(list(free_cells.difference(map_.base)))
+
+    def move_to_priority(self, map_: GameMap, state: GameState
+                         ) -> Optional[tuple[Actions, dict]]:
+        """
+        Method finds path to priority, and call move action
+        with reachable empty cell on the path to priority,
+        according to current game situation
+        :param map_: GameMap obj
+        :param state: GameState obj
+        :return:
+        """
+        step_cell = None
+        path_to_priority = self.model.coordinates.a_star(map_.available_cells,  self.priority)
+        speed_points = self.speed
+        if speed_points >= len(path_to_priority):
+            step_cell = path_to_priority[-1]
+        else:
+            while speed_points:
+                if path_to_priority[speed_points - 1] in state.tank_cells:
+                    speed_points -= 1
+                else:
+                    step_cell = path_to_priority[speed_points - 1]
+                    break
+        if step_cell and step_cell not in state.tank_cells:
+            return self.move(step_cell)
+        return None
+
+    @staticmethod
+    def build(t_id: int, spec: TankModel) -> "Vehicle":
+        """
+        fabric method instantiates vehicles of given spec
+        :param t_id: tanks id
+        :param spec: TankModel obj
+        :return: Vehicle of proper type
+        """
+        match spec.vehicle_type:
+            case "medium_tank":
+                return MediumTank(t_id, spec)
+            case "light_tank":
+                return LightTank(t_id, spec)
+            case "heavy_tank":
+                return HeavyTank(t_id, spec)
+            case "at_spg":
+                return AtSpg(t_id, spec)
+            case "spg":
+                return Spg(t_id, spec)
+
+
+class MediumTank(Vehicle):
+    """
+    Medium tanks class that implements
+    specific strategy for medium tanks
+    """
+
+
+class LightTank(Vehicle):
+    """
+    Light tanks class that implements
+    specific strategy for light tanks
+    """
+
+
+class HeavyTank(Vehicle):
+    """
+    Heavy tanks class that implements
+    specific strategy for heavy tanks
+    """
+
+
+class AtSpg(Vehicle):
+    """
+    AtSpg class that implements
+    specific strategy for AtSpg
+    """
+
+    def targets_in_range(self, state: GameState, map_: GameMap
+                         ) -> Union[set, set[tuple[int, int, int]]]:
```

Comment:
Why do you use Union here? 
Also it is a good practice to use type aliases for difficult structures with couple of levels of nesting. For example:
`T_COORDINATE = Tuple[int, int, int] `
Than you may use this alias in your type hints. Like this:
`Set[T_COORDINATE]`